### DISCUSSION TITLE
How do I get the latent gradient respect to my model

### DISCUSSION THREAD

**Discussion by Rami-Ismael** [larq/larq#754] [Q&A]
My project require me to track the gradient of the model.

**Comment by Tombana** [comment]
Could you elaborate, and share some code for what you are trying to do?

**Reply by Rami-Ismael** [reply]
Before asking the question, I did not read the Latent Weights Do Not Exist: Rethinking Binarized Neural Network Optimization. The goal is to introduce fault to the weight of a BNN. Then what is the effect on the gradient of the model? Am I correct to believe there is no latent gradient-like for latent weight. The only gradient the Larq use is the pseudo-gradient. 

``` python
  for current_model_index in range(0 , number_of_category):
    new_model = copy.deepcopy(model)
    ## select a model index that will flip
    random_index = random.randint(0 , number_of_category)
    ##get the number of one and negative on in the model
    number_of_ones_in_the_model = number_of_ones(model)
    number_of_negative_ones = number_of_negatives_ones(model)
    layer_name  = flip_index(model = model , index = random_index , verbose = False)
    if layer_name ==None:
      ValueError(f"The index {current_model_index} dose not exist in a layer")
    else:
      print(f"The current Step is {current_model_index}" )

    ## get the current numbe of ones and negative after the bit flips 
    new_number_number_of_ones_in_the_model = number_of_ones(model)
    new_number_of_negatives_ones = number_of_negatives_ones(model)
    ## this will prove only one bit flip occur
    assert abs(number_of_ones_in_the_model-new_number_number_of_ones_in_the_model)==1
    assert abs(number_of_negative_ones-new_number_of_negatives_ones)==1
    for image_batch , label_batch in ds:
      with tf.GradientTape() as tape:
        with larq.context.quantized_scope(True):
          prediction  = new_model(image_batch)
          loss = new_model.loss(label_batch , prediction)
          gradients = tape.gradient(loss, new_model.trainable_variables)
        ## Gradients with respect to a model
      assert len(gradients) >0
      assert None not in gradients 
      ## loop through each gradient
      for gradient_index in range(len(gradients)):
        ## add the index of index where the bit was flip
        #list_train_gradients[train_idx] = gradients[index].numpy().flatten()
        new_element = gradients[gradient_index].numpy().flatten()
        old_column = new_element.shape[0]
        new_element = new_element.reshape(1 ,new_element.shape[0] )
        ## append the number of fault  in the matrix 
        new_element = np.append(new_element , [current_model_index])
        new_element = new_element.reshape(1 ,new_element.shape[0] )
        assert new_element.shape[1]==old_column+1
        ## add the name of layer where the bit is flip
        new_element = np.append(new_element , [layer_name])
        new_element = new_element.reshape(1 ,new_element.shape[0] )
        assert new_element.shape[1]==old_column+2
``

**Reply by Tombana** [reply]
> Am I correct to believe there is no latent gradient-like for latent weight. The only gradient the Larq use is the pseudo-gradient.

Could you elaborate on this? On the one hand there is the derivative of the loss with respect to the binary weight, and on the other hand the derivative with respect to the latent weight. The latter is either zero (the true derivative) or the same as the former when using the straight-through-estimator.

**Reply by Rami-Ismael** [reply]
In Larq, there is easy access to the latent weight and the binary weight. When the BNN trains the model, I wondered if it uses the gradient with respect to the model latent weight or binary weight. Second,  the gradient I got from the loss respect to the trainable weight variable is not zero

**Reply by jneeven** [reply]
@Rami-Ismael when training BNNs, we typically use the [Straight-Through Estimator (STE)](https://docs.larq.dev/larq/api/quantizers/#stesign). As you can see in the figures there, the gradient of this quantizer is simply 1 wherever the inputs are between [-1, 1] and 0 otherwise. In other words: we pretend we never did any quantization at all, and then just clip the gradient in that range.

>  When the BNN trains the model, I wondered if it uses the gradient with respect to the model latent weight or binary weight.

The answer here would therefore be the latent weight, since the gradients don't "know" the weights were quantized at any point. However, because the gradients do get clipped when the weights become larger than 1 or smaller than -1, the latent weights should still roughly be in that range.

**Reply by Rami-Ismael** [reply]
Am I correct to say. There is no way to get the gradient respect to the  binary weights.

**Reply by jneeven** [reply]
> Am I correct to say. There is no way to get the gradient respect to the binary weights.

Correct, because no such gradient exists :+1:
